README.txt
VenezuelanEidAuthPlugin Overview

 VenezuelanEidAuthPlugin is a PAS Plugin that gives the user the possibility to identify himself using is Venezuelan electronic identity card. 

 This plugin can be used together with other PAS Plugins.

 This product should be installed together with CPDescriptive that extends users descriptive. 

 It will automatically add a "nationalregister" property to the portal_memberdata tool. If you do not want to use CPDescriptive, make sure a "nationalregister" property exists in portal_memberdata or it will not be possible to connect to the Plone Site using this plugin.

 A correctly configured web server (Apache, IIS, ...) has to be configured to accept HTTPS and SSL certificates. See the documentation on www.communesplone.org.

 After installing the product, define the https address the connection will have to redirect to in ZMI -> portal_properties, fill the https_address property.


HISTORY.txt
20060725 0.1 Beta1
--> Tests have been writed !!!
--> Some modifications and bugs have been corrected during the tests ;-)
--> Added support for logged_in.cpy script
--> More inserted in Plone through templates customization

200607010.1 Alpha2
--> Connection based on national register number :
        --> connection is available manually (username+passwd) and with eID card
        --> connect user with eID if subscribed (lookup national register number)
--> Depends on CPDescriptive product for memberdata extension and subscribing form

20060615 0.1 Alpha1
--> Connect with an existing user, without testing


tests/__init__.py
# -*- coding: utf-8 -*-
#
# File: __init__.py
#
# Copyright (c) 2006 by CommunesPlone
#
# GNU General Public License (GPL)
#
# This program is free software; you can redistribute it and/or
# modify it under the terms of the GNU General Public License
# as published by the Free Software Foundation; either version 2
# of the License, or (at your option) any later version.
#
# This program is distributed in the hope that it will be useful,
# but WITHOUT ANY WARRANTY; without even the implied warranty of
# MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE. See the
# GNU General Public License for more details.
#
# You should have received a copy of the GNU General Public License
# along with this program; if not, write to the Free Software
# Foundation, Inc., 51 Franklin Street, Fifth Floor, Boston, MA
# 02110-1301, USA.
#

__author__ = """Gauthier BASTIEN <gbastien@commune.sambreville.be>"""
__docformat__ = 'plaintext'


##code-section init-module-header #fill in your manual code here
##/code-section init-module-header


# Subpackages

# Classes

##code-section init-module-footer #fill in your manual code here
##/code-section init-module-footer



tests/testAllVenezuelanEidAuthPlugin.py
# -*- coding: utf-8 -*-
#
# File: testAllVenezuelanEidAuthPlugin.py
#
# Copyright (c) 2006 by CommunesPlone
#
# GNU General Public License (GPL)
#
# This program is free software; you can redistribute it and/or
# modify it under the terms of the GNU General Public License
# as published by the Free Software Foundation; either version 2
# of the License, or (at your option) any later version.
#
# This program is distributed in the hope that it will be useful,
# but WITHOUT ANY WARRANTY; without even the implied warranty of
# MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE. See the
# GNU General Public License for more details.
#
# You should have received a copy of the GNU General Public License
# along with this program; if not, write to the Free Software
# Foundation, Inc., 51 Franklin Street, Fifth Floor, Boston, MA
# 02110-1301, USA.
#

__author__ = """Gauthier BASTIEN <gbastien@commune.sambreville.be>"""
__docformat__ = 'plaintext'

import os, sys
if __name__ == '__main__':
    execfile(os.path.join(sys.path[0], 'framework.py'))

##code-section module-header #fill in your manual code here
from AccessControl import Unauthorized

from AccessControl.SecurityManagement import getSecurityManager

from Products.CMFCore.permissions import View
from Products.CMFCore.utils import getToolByName

##/code-section module-header

#
# Test-cases for class(es) 
#

from Testing import ZopeTestCase
from Products.VenezuelanEidAuthPlugin.config import *
from Products.VenezuelanEidAuthPlugin.tests.BaseVenezuelanEidAuthPluginTestCase import BaseVenezuelanEidAuthPluginTestCase

# Import the tested classes

##code-section module-beforeclass #fill in your manual code here
##/code-section module-beforeclass


class testAllVenezuelanEidAuthPlugin(BaseVenezuelanEidAuthPluginTestCase):
    """Test-cases for class(es) ."""

    ##code-section class-header_testAllVenezuelanEidAuthPlugin #fill in your manual code here
    ##/code-section class-header_testAllVenezuelanEidAuthPlugin

    def afterSetUp(self):
        """

        """
        BaseVenezuelanEidAuthPluginTestCase.afterSetup(self)

    # Manually created methods
    def testExtensionsInstall(self):
        #we test if the property called https_address is added into site_properties
        self.failUnless(hasattr(self.portal.portal_properties.site_properties, 'https_address'))

        #we test if the MemberWithEid role exist in PAS
        pas = self.portal.acl_users
        prm = pas.portal_role_manager
        self.failUnless( "MemberWithEid" in prm.listRoleIds())

        #we test if the VenezuelanEidAuthPlugin plugin is added in plonePas
        self.failUnless( hasattr(pas, 'VenezuelanEidAuthPlugin'))

    def testExtractCredentials(self):
        """
         We test if the extractCredentials method do what it has to...
        """
        portal = self.portal
        REQUEST = portal.REQUEST
        SESSION = portal.REQUEST.SESSION
        #we add the plugin
        #portal.acl_users.manage_addProduct['VenezuelanEidAuthPlugin'].manage_addVenezuelanEidAuthPlugin(id='beap')
        beap = portal.acl_users.VenezuelanEidAuthPlugin
        #--------------------> WITHOUT HTTPS <--------------------
        #we check that extractCredentials return None if the REQUEST does not have 'HTTP_SSL_CLIENT_S_DN'
        #weird behaviour of REQUEST that has got every keys ???  We str(REQUEST)...
        self.failIf('HTTP_SSL_CLIENT_S_DN' in str(REQUEST))
        self.failIf(beap.extractCredentials(REQUEST))
        
        #moreover, the SESSION object has not the keys 'eid_nr' and 'eid_from_http'
        self.failIf(SESSION.has_key('eid_nr'))
        self.failIf(SESSION.has_key('eid_from_http'))
        self.failIf(beap.extractCredentials(REQUEST))
        
        #--------------------> WITH HTTPS <--------------------
        #we set 'HTTP_SSL_CLIENT_S_DN' to "/C=VE/CN=User Venezuelan (Authentication)/SN=Venezuelan/GN=User/serialNumber=71715100070"
        REQUEST.set('HTTP_SSL_CLIENT_S_DN', "/C=VE/CN=User Venezuelan (Authentication)/SN=Venezuelan/GN=User/serialNumber=71715100070")
        self.failUnless('HTTP_SSL_CLIENT_S_DN' in str(REQUEST))
        
        #extractCredentials will return the creds
        creds = ({'eid_nr': '71715100070',
                  'eid_from_http': 1,
                  'eid_http_ssl_client_s_dn': "/C=VE/CN=User Venezuelan (Authentication)/SN=Venezuelan/GN=User/serialNumber=71715100070"
                 })
        #first time we extractCredentials, the credentials are set in SESSION
        self.assertEquals(beap.extractCredentials(REQUEST), creds)
        self.failUnless(SESSION.has_key('eid_nr'))
        self.failUnless(SESSION.has_key('eid_from_http'))
        
        #when extractCredentials has returned the credentials once, it will always return them
        self.assertEquals(beap.extractCredentials(REQUEST), creds)

        #we check if we change the 'HTTP_SSL_CLIENT_S_DN' if the user is no more recongnized and the process of authentication start again
        old_creds = beap.extractCredentials(REQUEST)
        REQUEST.set('HTTP_SSL_CLIENT_S_DN', "/C=VE/CN=User2 Venezuelan2 (Authentication)/SN=Venezuelan2/GN=User2/serialNumber=71715100070")
        self.assertNotEquals(beap.extractCredentials(REQUEST), old_creds)
        
        #check if Apache remove the 'HTTP_SSL_CLIENT_S_DN' from the REQUEST
        REQUEST.set('HTTP_SSL_CLIENT_S_DN', None)
        self.failIf(beap.extractCredentials(REQUEST))
        #when Apache has just removed 'HTTP_SSL_CLIENT_S_DN' from the REQUEST, we must be sure that extractCredentials has cleaned up SESSION : keys 'eid_from_http', 'eid_nr', 'eid_http_ssl_client_s_dn' and 'eid_username' must have been removed
        self.failIf(SESSION.has_key('eid_nr'))
        self.failIf(SESSION.has_key('eid_from_http'))
        self.failIf(SESSION.has_key('eid_http_ssl_client_s_dn'))
        self.failIf(SESSION.has_key('eid_username'))
        #we call extractCredentials again to test the try with del request.SESSION
        self.failIf(beap.extractCredentials(REQUEST))


    def testGetClientData(self):
        """
         We test the getClientData() method that returns the credentials from 'HTTP_SSL_CLIENT_S_DN'
        """
        portal = self.portal
        #we add the plugin
        #portal.acl_users.manage_addProduct['VenezuelanEidAuthPlugin'].manage_addVenezuelanEidAuthPlugin(id='beap')
        beap = portal.acl_users.VenezuelanEidAuthPlugin
        
        from_http = None
        self.failIf(beap.getClientData(from_http))
        
        #we set the from_http with some wrong values and we test
        #not valid
        from_http = "1"
        self.failIf(beap.getClientData(from_http))
        #not valid
        from_http = "12345678910"
        self.failIf(beap.getClientData(from_http))
        #not a valid serialNumber
        from_http = "/C=VE/CN=User Venezuelan (Authentication)/SN=Venezuelan/GN=User/serialNumber=1234567891"
        self.failIf(beap.getClientData(from_http))
        
        #we check that a valid 'HTTP_SSL_CLIENT_S_DN' can be evaluated to return the serialNumber
        from_http = "/C=VE/CN=User Venezuelan (Authentication)/SN=Venezuelan/GN=User/serialNumber=71715100070"
        self.assertEquals(beap.getClientData(from_http), "71715100070")


    def testAuthenticateCredentials(self):
        """
         We test if the authenticateCredentials method do what it has to...
        """
        portal = self.portal
        REQUEST = portal.REQUEST
        SESSION = portal.REQUEST.SESSION
        #we add the plugin
        #portal.acl_users.manage_addProduct['VenezuelanEidAuthPlugin'].manage_addVenezuelanEidAuthPlugin(id='beap')
        beap = portal.acl_users.VenezuelanEidAuthPlugin
        
        #--------------------> CREDENTIALS WITHOUT 'eid_from_http' KEY <--------------------
        #if the credentials returned by extractCredentials is None, authenticateCredentials must return None too...
        credentials = {}
        self.failIf(beap.authenticateCredentials(credentials))
        
        #if the credentials returned by extractCredentials do not have the key 'eid_from_http', authenticateCredentials return None
        credentials = ({'eid_nr': '71715100070'})
        self.failIf(beap.authenticateCredentials(credentials))
        
        #--------------------> CREDENTIALS WITH 'eid_from_http' KEY <--------------------
        #--------------------> USER DOES NOT EXIST <--------------------
        #credentials is well formatted
        #but has not the 'eid_username' key
        credentials = ({'eid_nr': '71715100070',
                  'eid_from_http': 1,
                  'eid_http_ssl_client_s_dn': "/C=VE/CN=User Venezuelan (Authentication)/SN=Venezuelan/GN=User/serialNumber=71715100070"
                 })
        #authenticateCredentials will look up the user
        #if the user does not exist, authenticateCredentials return None
        self.failIf(beap.authenticateCredentials(credentials))
        #and 'eid_username' is not set in SESSION
        self.failIf(SESSION.get('eid_username'))

        #--------------------> USER EXISTS <--------------------
        #now, we add the user in acl_users and we try to authenticateCredentials...
        #in order to search the user upon his national register number, and so for the 'getUserNameFromNR' to work, we need to add a property to portal_memberdata called 'nationalregister' and to fill it.  This can be done in Plone using the CPDescriptive product...
        portal.portal_memberdata.manage_addProperty('nationalregister', '', 'string')
        #addMember need id, password, and if we enter some properties, we have to specify the username and email too...
        pr_tool = getToolByName(portal, 'portal_registration')
        pr_tool.addMember('user', '12345', properties={'username':'User',
                                                       'email':'user@user.be',
                                                       'nationalregister':'71715100070'})
        #the user exist, authenticateCredentials must return this user
        self.assertEquals(beap.authenticateCredentials(credentials), ('user', 'user'))
        #if the user has been found, authenticateCredentials set 'eid_username' is SESSION
        self.assertEquals(SESSION.get('eid_username'), "user")
        
    def testGetUserNameFromNR(self):
        """
         We test that getUserNameFromNR work correctly
        """
        portal = self.portal
        REQUEST = portal.REQUEST
        SESSION = portal.REQUEST.SESSION
        #we add the plugin
        #portal.acl_users.manage_addProduct['VenezuelanEidAuthPlugin'].manage_addVenezuelanEidAuthPlugin(id='beap')
        beap = portal.acl_users.VenezuelanEidAuthPlugin

        nr = "71715100070"
                
        #--------------------> USER DOES NOT EXIST <--------------------
        #if the user does not exists, the method obviously return None but it work even if the 'nationalregister' property does not exist in portal_memberdata
        self.failIf(beap.getUserNameFromNR(nr))
        
        #--------------------> USER EXISTS <--------------------
        #we add the 'nationalregister' porperty in portal_memberdata
        portal.portal_memberdata.manage_addProperty('nationalregister', '', 'string')
        pr_tool = getToolByName(portal, 'portal_registration')
        #we add a user with a valid 'nationalregister'
        #addMember need id, password, and if we enter some properties, we have to specify the username and email too...
        pr_tool.addMember('user', '12345', properties={'username':'User',
                                                       'email':'user@user.be',
                                                       'nationalregister':'71715100070'})
        #now it should work...
        self.assertEquals(beap.getUserNameFromNR(nr), "user")


    def testExtractAndAuthenticateCredentials(self):
        """
         As extractCredentials and authenticateCredentials are linked, we have to test them together too...
         What we have to test here is what is modified in extractCredentials and used in authenticateCredentials : the 'eid_username' key from SESSION
        """
        portal = self.portal
        REQUEST = portal.REQUEST
        SESSION = portal.REQUEST.SESSION
        #we add the plugin
        #portal.acl_users.manage_addProduct['VenezuelanEidAuthPlugin'].manage_addVenezuelanEidAuthPlugin(id='beap')
        beap = portal.acl_users.VenezuelanEidAuthPlugin
        #we add the 'nationalregister' porperty in portal_memberdata
        portal.portal_memberdata.manage_addProperty('nationalregister', '', 'string')
        pr_tool = getToolByName(portal, 'portal_registration')
        #we add a user with a valid 'nationalregister'
        #addMember need id, password, and if we enter some properties, we have to specify the username and email too...
        pr_tool.addMember('user', '12345', properties={'username':'User',
                                                       'email':'user@user.be',
                                                       'nationalregister':'71715100070'})

        REQUEST.set('HTTP_SSL_CLIENT_S_DN', "/C=VE/CN=User Venezuelan (Authentication)/SN=Venezuelan/GN=User/serialNumber=71715100070")
        #we extract credentials ...
        creds = beap.extractCredentials(REQUEST)
        #... and the user must have been authenticated
        self.assertEquals(beap.authenticateCredentials(creds), ('user', 'user'))
        #the 'eid_username' key must be set in the SESSION
        self.failUnless(SESSION.has_key('eid_username'))
        
        #if we change the 'HTTP_SSL_CLIENT_S_DN' (as Apache could do), the 'eid_username' key must have disappeared after extractCredentials but must be set after authenticateCredentials if the serialNumber is found in the users 'nr'
        REQUEST.set('HTTP_SSL_CLIENT_S_DN', "/C=VE/CN=User2 Venezuelan2 (Authentication)/SN=Venezuelan/GN=User/serialNumber=71715100070")
        
        creds = beap.extractCredentials(REQUEST)
        self.failIf(SESSION.has_key('eid_username'))
        beap.authenticateCredentials(creds)
        self.failUnless(SESSION.has_key('eid_username'))
        
        #if we change the 'HTTP_SSL_CLIENT_S_DN' AND the serialNumber is not found in the users 'nr', the 'eid_username' key can not be set in SESSION after extractCredentials and after authenticateCredentials
        REQUEST.set('HTTP_SSL_CLIENT_S_DN', "/C=VE/CN=User2 Venezuelan2 (Authentication)/SN=Venezuelan/GN=User/serialNumber=71715100169")
        creds = beap.extractCredentials(REQUEST)
        self.failIf(SESSION.has_key('eid_username'))
        beap.authenticateCredentials(creds)
        self.failIf(SESSION.has_key('eid_username'))
        

    def testLoggedInScriptExecutedWithNotRedirectableURL(self):
        """
         We want the logged_in.cpy script to be executed each time we connect (and only once)
        """
        portal = self.portal
        REQUEST = portal.REQUEST
        SESSION = portal.REQUEST.SESSION
        #we add the plugin
        #portal.acl_users.manage_addProduct['VenezuelanEidAuthPlugin'].manage_addVenezuelanEidAuthPlugin(id='beap')
        beap = portal.acl_users.VenezuelanEidAuthPlugin
        #we add the 'nationalregister' porperty in portal_memberdata
        portal.portal_memberdata.manage_addProperty('nationalregister', '', 'string')
        pr_tool = getToolByName(portal, 'portal_registration')
        #we add a user with a valid 'nationalregister'
        #addMember need id, password, and if we enter some properties, we have to specify the username and email too...
        pr_tool.addMember('user', '12345', properties={'username':'User',
                                                       'email':'user@user.be',
                                                       'nationalregister':'71715100070'})

        REQUEST.set('HTTP_SSL_CLIENT_S_DN', "/C=VE/CN=User Venezuelan (Authentication)/SN=Venezuelan/GN=User/serialNumber=71715100070")
        #'logged_out' will not be redirected by login_next
        REQUEST.set('URL', "logged_out")
        
        creds = beap.extractCredentials(REQUEST)
        #we verify that the logged_in.cpy script is executed if we are not using a not_redirectable_url
        self.failIf(SESSION.has_key('eid_logged_in_executed'))
        beap.authenticateCredentials(creds)
        self.failIf(SESSION.has_key('eid_logged_in_executed'))
        
        #we put a redirectable URL in REQUEST/'URL'
        REQUEST.set('URL', "redirectable_url")
        beap.authenticateCredentials(creds)
        self.failUnless(SESSION.has_key('eid_logged_in_executed'))
        
        
    def testLoggedInScriptExecutedWithRedirectableURL(self):
        """
         We want the logged_in.cpy script to be executed each time we connect (and only once)
        """
        portal = self.portal
        REQUEST = portal.REQUEST
        SESSION = portal.REQUEST.SESSION
        #we add the plugin
        #portal.acl_users.manage_addProduct['VenezuelanEidAuthPlugin'].manage_addVenezuelanEidAuthPlugin(id='beap')
        beap = portal.acl_users.VenezuelanEidAuthPlugin
        #we add the 'nationalregister' porperty in portal_memberdata
        portal.portal_memberdata.manage_addProperty('nationalregister', '', 'string')
        pr_tool = getToolByName(portal, 'portal_registration')
        #we add a user with a valid 'nationalregister'
        #addMember need id, password, and if we enter some properties, we have to specify the username and email too...
        pr_tool.addMember('user', '12345', properties={'username':'User',
                                                       'email':'user@user.be',
                                                       'nationalregister':'71715100070'})

        REQUEST.set('HTTP_SSL_CLIENT_S_DN', "/C=VE/CN=User Venezuelan (Authentication)/SN=Venezuelan/GN=User/serialNumber=71715100070")
        
        creds = beap.extractCredentials(REQUEST)
        #we verify that the logged_in.cpy script is executed if we are not using a not_redirectable_url
        self.failIf(SESSION.has_key('eid_logged_in_executed'))
        beap.authenticateCredentials(creds)
        self.failUnless(SESSION.has_key('eid_logged_in_executed'))
        

def test_suite():
    from unittest import TestSuite, makeSuite
    suite = TestSuite()
    suite.addTest(makeSuite(testAllVenezuelanEidAuthPlugin))
    return suite

##code-section module-footer #fill in your manual code here
##/code-section module-footer

if __name__ == '__main__':
    framework()




VenezuelanEidAuthPlugin.py
# -*- coding: utf-8 -*-
#
# File: VenezuelanEidAuthPlugin
#
# Copyright (c) 2006 by CommunesPlone
#
# GNU General Public License (GPL)
#
# This program is free software; you can redistribute it and/or
# modify it under the terms of the GNU General Public License
# as published by the Free Software Foundation; either version 2
# of the License, or (at your option) any later version.
#
# This program is distributed in the hope that it will be useful,
# but WITHOUT ANY WARRANTY; without even the implied warranty of
# MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE. See the
# GNU General Public License for more details.
#
# You should have received a copy of the GNU General Public License
# along with this program; if not, write to the Free Software
# Foundation, Inc., 51 Franklin Street, Fifth Floor, Boston, MA
# 02110-1301, USA.
#

__author__ = """Gauthier BASTIEN <gbastien@commune.sambreville.be>"""
__docformat__ = 'plaintext'

from AccessControl import ClassSecurityInfo

from Globals import InitializeClass
from OFS.Cache import Cacheable

from Products.PluggableAuthService.plugins.BasePlugin import BasePlugin
from Products.PluggableAuthService.utils import classImplements
from Products.PluggableAuthService.interfaces.plugins import IAuthenticationPlugin
from Products.PluggableAuthService.interfaces.plugins import IExtractionPlugin

import logging
#from zLOG import LOG, PROBLEM
from Products.PageTemplates.PageTemplateFile import PageTemplateFile

from Products.CMFCore.utils import getToolByName

manage_addVenezuelanEidAuthPluginForm = PageTemplateFile(
    'www/VenezuelanEidAdd', globals(), __name__='manage_addVenezuelanEidAuthPluginForm' )

def manage_addVenezuelanEidAuthPlugin(dispatcher, id, title=None, REQUEST=None):
    """ Add a VenezuelanEidAuthPlugin to a Pluggable Auth Service. """

    obj = VenezuelanEidAuthPlugin(id, title)
    
    dispatcher._setObject(obj.getId(), obj)

    if REQUEST is not None:
        REQUEST['RESPONSE'].redirect(
                                '%s/manage_workspace'
                                '?manage_tabs_message='
                                'VenezuelanEidAuthPlugin+added.'
                            % dispatcher.absolute_url())

class VenezuelanEidAuthPlugin(BasePlugin, Cacheable):
    """ PAS plugin for using VenezuelanEid credentials to log in. """

    meta_type = 'VenezuelanEidAuthPlugin'

    security = ClassSecurityInfo()
    
    def __init__(self, id, title=None):
        self._id = self.id = id
        self.title = title

    #
    #   IAuthenticationPlugin implementation
    #
    security.declarePrivate('authenticateCredentials')
    def authenticateCredentials(self, credentials):

        """ 
            See IAuthenticationPlugin.
            o We expect the credentials to be those returned by ILoginPasswordExtractionPlugin.
            o We do not need a password if we receive can access data in the REQUEST
            
            XXX We should not use the lookup if we receive valid HTTPS credentials and the user has not been found once, because the user could connect using his username/passwd and we could check to many times
        """
        #print "credentials = ", credentials
        debug = False
        #print "VenezuelanEidAuthPlugin : debug mode is %s" % debug
        
        if debug:
            return "User", "user"
            
        if credentials.has_key('eid_from_http'):
            #we received something, the user is using his eID card, proceed
            
            if self.REQUEST.SESSION.has_key('eid_username'):
                #we already check in users if the actual user exist, we return it
                #we return the user
                #print "return the already connected user"
                #if we come from logged_out, login_success, ... we need to call the logged_in script
                #these URL are not redirected by login_next
                #see login_next.cpy from CMFPlone for more informations
                if not self.REQUEST.SESSION.has_key('eid_logged_in_executed'):
                    #print "executed in first if"
                    self.REQUEST.SESSION.set('eid_logged_in_executed', 1)
                    portal = getToolByName(self, 'portal_url').getPortalObject()
                    self.REQUEST.RESPONSE.redirect(portal.portal_properties.site_properties.getProperty('https_address') + 'logged_in?came_from=%s' % self.REQUEST.get('URL'))
                return self.REQUEST.SESSION.get('eid_username'), self.REQUEST.SESSION.get('eid_username')
            else:
                #lookup user national register in registered users
                #print "lookup user"
                user_name = self.getUserNameFromNR(credentials['eid_nr'])
                if user_name:
                    #we force the redirection to logged_in to be sure that logged_in.cpy where we manage the MemberWithEid role is executed
                    #we do not call logged_in here if the url is not redirectable (see login_next.cpy in CMFPlone)
                    if not self.REQUEST.SESSION.has_key('eid_logged_in_executed'):
                        not_redirectable_urls = ['login', 'login_success', 'login_password', 'login_failed',
                                               'logged_in', 'logged_out', 'logout', 'registered',
                                               'mail_password', 'mail_password_form', 'join_form',
                                               'require_login', 'member_search_results'] 
                        redirectable = True
                        for not_redirectable_url in not_redirectable_urls:
                            if not_redirectable_url in self.REQUEST.get('URL'):
                                #print "not redir found"
                                redirectable = False
                                break
                                
                        if redirectable:
                            #print "redir found and executed"
                            self.REQUEST.SESSION.set('eid_logged_in_executed', 1)
                            portal = getToolByName(self, 'portal_url').getPortalObject()
                            self.REQUEST.RESPONSE.redirect(portal.portal_properties.site_properties.getProperty('https_address') + 'logged_in?came_from=%s' % self.REQUEST.get('URL'))
                    
                    self.REQUEST.SESSION.set('eid_username', user_name)
                    return user_name, user_name                    
                
                #we will return None if the user has not be found in the database
                return None
        else:
            #if credentials is not None, it means that extractCredentials returned something without the 'eid_from_http' key, it should not happen...
            #print "return None"
            return None


    security.declarePrivate('extractCredentials')
    def extractCredentials(self, request):
        """ 
            Extract eid userinfo from request 
            These informations will be used by authenticateCredentials as it receive them as parameter
        """
        #if the user try to connect using his eID card and that the getClientData has returned the 'eid_nr'
        #we could suppose that we have eid_from_http but not 'eid_nr' altought it should not happen...
        
        creds = {}
        #we know that we have already extracted the credentials only if :
        #--> 'eid_from_http' is set in SESSION --> we have already encountered the HTTPS support
        #--> 'eid_nr' is set in SESSION --> we have successfully extracted nr from 'HTTP_SSL_CLIENT_S_DN' with the getClientData method
        #--> 'HTTP_SSL_CLIENT_S_DN' is in the REQUEST --> Apache still send the SSL var
        #--> 'HTTP_SSL_CLIENT_S_DN' in REQUEST is equal to 'eid_http_ssl_client_s_dn' in SESSION wich means that the user as not changed
        
        if request.SESSION.has_key('eid_from_http') and request.get('HTTP_SSL_CLIENT_S_DN') and request.get('HTTP_SSL_CLIENT_S_DN') == request.SESSION.get('eid_http_ssl_client_s_dn'):
                #we still check if 'HTTP_SSL_CLIENT_S_DN' is in the REQUEST because Apache could remove it
                #we already parsed 'HTTP_SSL_CLIENT_S_DN', we use 'eid_nr' stored in SESSION object
                #as we just manage cached vars in SESSION, we do not need delete 'eid_username' from SESSION
                creds.update({'eid_nr':request.SESSION.get('eid_nr'),
                            'eid_from_http':1,
                            'eid_http_ssl_client_s_dn': request.get('HTTP_SSL_CLIENT_S_DN')})
                return creds
        else:
            #we play with 'HTTP_SSL_CLIENT_S_DN'
            from_http = request.get('HTTP_SSL_CLIENT_S_DN')
            if from_http:
                #we search the national register number in the request send from Apache
                ret = self.getClientData(from_http)
                if ret:
                    (nr, last_name, first_name, full_name) = ret
                    creds.update({'eid_nr':nr,
                                  'eid_from_http':1,
                                  'eid_http_ssl_client_s_dn': from_http
                                })
                    request.SESSION.set('eid_nr', creds['eid_nr'])
                    request.SESSION.set('eid_last_name', last_name)
                    request.SESSION.set('eid_first_name', first_name)
                    #request.SESSION.set('eid_full_name', full_name)
                    request.SESSION.set('eid_from_http', creds['eid_from_http'])
                    #we save the 'HTTP_SSL_CLIENT_S_DN' in the SESSION to see if it is always the same send by Apache
                    request.SESSION.set('eid_http_ssl_client_s_dn', creds['eid_http_ssl_client_s_dn'])
                    #we try to remove 'eid_username' from SESSION to force AuthenticateCredentials to do the authentication again if it had already been done
                    try:
                        del request.SESSION['eid_username']
                    except KeyError:
                        pass
                    return creds
                else:
                    creds.update({'eid_from_http':0})
            else:
                creds.update({'eid_from_http':0})
                    
        #If we can not get this from the REQUEST, we are not in a correctly configured HTTPS mode
        #we remove the keys from SESSION !!!
        #--> 'eid_from_http' in creds == 0 if from_hhtp is None or if nr returned by getClientData is None
        
        #XXX performance problem at trying to delete the keys from SESSION when not using eID card???  The plugin will always do this try/catch...
        if creds['eid_from_http'] == 0:
            try:
                del request.SESSION['eid_nr']
                del request.SESSION['eid_from_http']
                del request.SESSION['eid_http_ssl_client_s_dn']
                del request.SESSION['eid_username']
            except KeyError:
                #if 'eid_nr' or 'eid_from_hhtp' or 'eid_http_ssl_client_s_dn' or 'eid_username' does not exist in SESSION, we pass, they already have been deleted
                pass
        
        return None
        

    security.declarePrivate('getClientData')
    def getClientData(self, from_http):
        """ 
        We receive a String, we need to parse it to return the name, first_name and nr (national register number)
        """
        #there can be UTF/ISO encoding problems with OpenSSL/Apache, so we do what we have to to correct this
        #UTF-8 codes are not passed as codes but as string
        #"é" should be '\xc3\xa9' but it is returned as '\\xc3\\xa9'
        nr = first_name = last_name = full_name = None
        
        try:
            #we need to correct the string if we wish to retrieve CN/SN/GN datas from from_http
            corrected_string = eval("u'" + from_http + "'")
            corrected_string = corrected_string.encode('latin1')
            corrected_string = unicode(corrected_string, 'utf-8')
            datas = corrected_string.split('/')
            #datas = from_http.split('/')
            #search for SN, GN and serialNumber
            for data in datas:
                if data[:12] == "serialNumber":
                    nr = data[13:]
                elif data[:3] == "SN=":
                    last_name = data[3:]
                elif data[:3] == "GN=":
                    first_name = data[3:]
                elif data[:3] == "CN=":
                    full_name = data[3:-17]

            #print "nr='%s',last='%s',first='%s',full='%s'"%(nr,last_name,first_name,full_name)
            #we check that there are 11 numbers left
            if len(nr) != 11:
                raise ValueError
    
            first_part = int(nr[0:9])
            last_part = int(nr[9:11])
    
            #the two last digits is the result of 97 les the modulo by 97 of the 10 first digits
            if last_part != (97 - (first_part%97)):
                raise ValueError
        except:
            #if we encoutered an error doing this, we have to stop here
            #we should not have errors here, so we LOG
            logger = logging.getLogger('VenezuelanEidAuthPlugin')
            logger.info("Failed to extract datas from from_http in getClientData(self, from_http).")
            return None
        
        return (nr, last_name, first_name, full_name)


    security.declarePrivate('getUserNameFromNR')
    def getUserNameFromNR(self, nr):
        """
            Lookup the user in the User Manager with is National Register
        """
        for user in self.acl_users.getUsers():
            props = self.acl_users.mutable_properties.getPropertiesForUser(user)._properties
            
            #the property 'nationalregister' could not exist, so we try and catch
            try:
                if props['nationalregister'] == nr:
                    return user.getId()
            except KeyError:
                pass
        
        return None


classImplements(VenezuelanEidAuthPlugin, IAuthenticationPlugin, IExtractionPlugin)

InitializeClass(VenezuelanEidAuthPlugin)